### PR TITLE
8305763 : Parsing a URI with an underscore goes through a silent exception, negatively impacting performance

### PR DESCRIPTION
**Pull request by dhanalla**
Issue 8305763 : Using underscores in the name for a URI triggers a silent exception in the java standard library, which consumes 5% of the CPU.

Exception:
java.net.URISyntaxException: Illegal character in hostname at index N: xyz1_abcd.com
    at java.base/java.net.URI$Parser.fail(URI.java:2943)
    at java.base/java.net.URI$Parser.parseHostname(URI.java:3487)
    at java.base/java.net.URI$Parser.parseServer(URI.java:3329)
    
    This exception is silent and does not produce any messages, except for ODP profiler, there is no other evidence that it’s happening (the stack trace above was printed after changes to Java library). The reason for this is because of how the URI creation is implemented in the java.net.URI class. There are two paths for creating a valid URI, and one of them goes through an exception.

We can see that if parseServer fails, there is still a way the authority gets assigned and we don’t throw an exception from the method. This means, not being able to parse the server is ok and the exception is silenced. In our case, the server parsing fails because we find an illegal character, as only alphanumeric and dash characters are allowed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8305763](https://bugs.openjdk.org/browse/JDK-8305763): Parsing a URI with an underscore goes through a silent exception, negatively impacting performance

### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13430/head:pull/13430` \
`$ git checkout pull/13430`

Update a local copy of the PR: \
`$ git checkout pull/13430` \
`$ git pull https://git.openjdk.org/jdk.git pull/13430/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13430`

View PR using the GUI difftool: \
`$ git pr show -t 13430`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13430.diff">https://git.openjdk.org/jdk/pull/13430.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13430#issuecomment-1504383491)

### CODE REVIEW

File: src/java.base/share/classes/java/net/URI.java [hunk 1]
```diff
@@ -3154,6 +3154,13 @@ private int scan(int start, int n, long lowMask, long highMask)
             return p;
         }
 
+        // parse as a Registry-based authority if not insisting
+        // upon a server-based authority
+        private boolean parseRegistryChars()
+        {
+            return requireServerAuthority?false:true;
+        }
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/net/URI.java:3162]: Ugh, this is very messy. Can you reformat this to be consistent with the existing code?

**Comment by dfuch** [comment]
From a quic look at the propose change, I got the feeling that this change might not be appropriate: I suspect it will let `host` be assigned to the reg_name.
We want to preserve the long standing behavior that:
```
jshell> new URI("http://foo_bar:8080/").getHost()
$1 ==> null
```
Is this still the case after your proposed changes?

**Comment by dhanalla** [comment]
Thanks for reviewing @dfuch,
With this change, the below operation returns hostname "foo_bar"

jshell> new URI("http://foo_bar:8080/").getHost()

I will work on addressing this to keep the old behavior unchanged.

**Comment by dhanalla** [comment]
Thanks @dfuch,  
I have validated the following tests with the new change.
1.  JShell:
       jshell> new URI("http://foo_bar:8080/").getHost()
       $1 ==> null

      jshell> new URI("http://foo_bar:8080/").getPort()
      $2 ==> -1
      
      jshell> new URI("http://foobar:8080/").getHost()
      $3 ==> "foobar"
      
      jshell> new URI("http://foobar:8080/").getPort()
      $4 ==> 8080
2.  all tests under "java/net/URI" passed as expected.
3. no failures related my change in tier1, tier2 and tier3 tests.

File: src/java.base/share/classes/java/net/URI.java [hunk 2]
```diff
@@ -3291,15 +3292,27 @@ private int parseAuthority(int start, int n)
...
+                    if (q < n) {
+                        if(skipParseException) {
+                            userInfo = null;
+                            host = null;
+                            port = -1;
+                            q = p;
+                        } else
+                            failExpecting("end of authority", q);
+                    } else
+                        authority = input.substring(p, n);
```
Comment (dfuch) [inline src/java.base/share/classes/java/net/URI.java:3314]:
```suggestion
                        } else {
                            failExpecting("end of authority", q);
                        }
                    } else {
                        authority = input.substring(p, n);
                    }
```

**Review by dfuch — COMMENTED** [review]
I have given this a round of testing and I believe the logic is now right.
I'd suggest adding curly braces around else blocks in case of nested if statements though (see suggestion below) as it makes it easier to verify which belongs to which.

File: src/java.base/share/classes/java/net/URI.java (hunk 2 above)
Comment (dhanalla) [reply src/java.base/share/classes/java/net/URI.java]: Thank you @dfuch

File: src/java.base/share/classes/java/net/URI.java [hunk 3]
```diff
@@ -3394,7 +3408,10 @@ private int parseServer(int start, int n)
                     }
                     p = q;
                 }
+            } else if( p < n && skipParseException) {
```
Comment (turbanoff) [inline src/java.base/share/classes/java/net/URI.java:3411]:
```suggestion
            } else if (p < n && skipParseException) {
```

File: src/java.base/share/classes/java/net/URI.java [hunk 4]
```diff
@@ -3524,8 +3541,13 @@ private int parseHostname(int start, int n)
             } while (p < n);
 
             if ((p < n) && !at(p, n, ':'))
+            {
+                if(skipParseException)
```
Comment (turbanoff) [inline src/java.base/share/classes/java/net/URI.java:3545]:
```suggestion
                if (skipParseException) {
```

File: src/java.base/share/classes/java/net/URI.java [hunk 5]
```diff
@@ -3291,15 +3292,28 @@ private int parseAuthority(int start, int n)
...
                 // to parse it as such.  If the attempt fails, try to treat it
                 // as a registry-based authority.
                 try {
-                    q = parseServer(p, n);
-                    if (q < n)
-                        failExpecting("end of authority", q);
-                    authority = input.substring(p, n);
+                    q = parseServer(p, n, skipParseException);
+                    if (q < n) {
+                        if(skipParseException) {
```
[the remark this answers is not in the document]
Comment (dhanalla) [reply src/java.base/share/classes/java/net/URI.java]: Thanks for reviewing the PR @turbanoff

**Pull request closed without merging**